src/plugins/intel_cpu/src/snippets_transformations/enforce_precision.cpp

```diff
@@ -0,0 +1,94 @@
+// Copyright (C) 2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "snippets_transformations/enforce_precision.hpp"
+
+#include <assert.h>
+#include <memory>
+
+#include "ov_ops/type_relaxed.hpp"
+#include "snippets/itt.hpp"
+#include "ngraph/rt_info.hpp"
+#include "snippets/pass/propagate_precision.hpp"
+
+using namespace ngraph;
+
+ov::intel_cpu::pass::EnforcePrecision::EnforcePrecision(
+    const element::Type source,
+    const element::Type target,
+    const std::shared_ptr<const ngraph::snippets::TargetMachine>& target_machine) : source(source), target(target), target_machine(target_machine) {
+}
+
+bool ov::intel_cpu::pass::EnforcePrecision::run_on_model(const std::shared_ptr<ov::Model>& f) {
+    RUN_ON_MODEL_SCOPE(PropagatePrecision);
+    OV_ITT_SCOPED_TASK(ngraph::pass::itt::domains::SnippetsTransform, "ov::intel_cpu::pass::EnforcePrecision")
+
+    bool was_updated = false;
+    auto& ops = f->get_ordered_ops();
+
+    for (auto op : ops) {
+        std::vector<size_t> input_precision_indexes;
+        for (size_t index = 0ull; index < op->get_input_size(); ++index) {
+            const auto& input = op->input(index);
+            const auto input_precision = input.get_source_output().get_element_type();
+            if (input_precision != source) {
+                continue;
+            }
+            input_precision_indexes.push_back(index);
+        }
+
+        if (input_precision_indexes.empty()) {
+            continue;
+        }
+
+        auto type_info = op->get_type_info();
+        OPENVINO_ASSERT(
+            target_machine->has(type_info),
+            "operation '" + std::string(type_info.version_id) + "::" + std::string(type_info.name) + "' was not found in target machine");
+
+        auto exec = target_machine->get_supported_precisions(type_info);
+        const auto supported_precisions = exec(op);
+
+        const auto insert_convert = [](
+            const Output<Node>& parent_output,
+            const std::shared_ptr<Node>& op,
+            const size_t input_index,
+            const element::Type& target) {
+            auto convert = std::make_shared<ngraph::snippets::op::ConvertSaturation>(
+                parent_output,
+                target);
+            ngraph::copy_runtime_info(parent_output.get_node_shared_ptr(), convert);
+            op->set_argument(input_index, convert);
+        };
+
+
+        bool need_to_validate = false;
+        if (supported_precisions.empty()) {
+            continue;
+        }
+
+        for (const auto& precisions : supported_precisions) {
+            for (auto index = 0; index < input_precision_indexes.size(); ++index) {
```

Comment:
I'm not sure that it's correct order of loops.
Because for example, we need convert `f32`->`bf16` and we have node with inputs `f32, f32, i32`. The corresponding emitter supports just `f32, f32, i32` and `bf16, bf16, i8` (just imagine).
Here we change first two inputs and leave last so: we get `bf16, bf16, i32`. But emitter doesn't support these precisions.
Could you please verify?

Reply:
not actual right now